### DISCUSSION TITLE
How to compare 2 JSON files with the same content but in a different order?

### DISCUSSION THREAD

**Discussion by jrchaves** [WinMerge/winmerge#2163] [Q&A] [answered]
Hi folks .... 

I have a specific situation which I have 2 JSON files that have the same content but in a different order:

1st JSON:

>         "Parameters": [
          {
            "HandleId": "0",
            "Data": {
              "$type": "ColorGrading",
              "brightness": 1,
              "contrast": 1,
             }
          },
          {
            "HandleId": "1",
            "Data": {
              "$type": "Vignette",
			}
		  },
        ]

2nd JSON file:

>         "Parameters": [
          {
            "HandleId": "1",
            "Data": {
              "$type": "Vignette",
			}
          },
          {
            "HandleId": "0",
            "Data": {
              "$type": "ColorGrading",
              "brightness": 1,
              "contrast": 1,
             }
		  },
        ]

As you can see, both JSON files have the same content but in a different order. I would like to compare them with WINMERGE but prior the comparison I would like to reorganize the 2nd JSON to be in the same order that the 1st JSON file, so there will be no difference detected between them.

Any idea will be very appreciated! 

best regards,

JR

**Comment by sdottaka** [comment]
WinMerge has an Unpacker plugin named QueryJSON using `jq`.
This plugin allows sorting.
However, the sample files are not valid JSON files with extra commas added at the end, so this plugin cannot process them.
So before sorting with QueryJSON, use the Replace plugin to remove unnecessary commas.

Specifically, specify the JSON files in the "Select Files or Folders" window and the following in the Unpacker Plugin as shown below, and press the [Compare] button.


~~~
Replace -e ",\s*\n\s*}" "}" -e ",\s*\n\s*\]" "]" | QueryJSON ".Parameters |= sort_by(.HandleId)"
~~~

![image](https://github.com/WinMerge/winmerge/assets/98126/4a524fef-6d58-4381-8411-eab1ec796133)
![image](https://github.com/WinMerge/winmerge/assets/98126/4819d1a1-9b4f-450e-903b-ea4797348c1e)

**Comment by jrchaves** [comment]
Hi Mr. Sawanaka,

Wow! I deeply thank you for providing such help! I didn´t know there was a QueryJSON plugin. Super thanks also for the detailed explanation. I just used the 2 examples in the way I did to simplify the question, but to be honest the files are different than the ones I shared. I tried to replicate the way you instructed me but for some reason I failed. So, I am attaching the "real" files and I appreciate if you could give an additional help with them.

As you can see, the HandleID's order is different (in the 1st FILE it begins with "HandleId": "0" and in the 2nd FILE it begins with "HandleId": "1", which avoid WinMerge to make an "usual" comparison. I would like that before the comparison, the order of the 2nd FILE be the same as the 1st FILE, I mean, moving (in the 2nd FILE) the "HandleId": "0" to the top  (to be in the same order than the 1st FILE), allowing the comparison between them.

The name of the files are different just to allows me to upload them and easily make reference to them (because to allow the comparison they should be the same).

Again, thank you so much for your time and support!

best regards,

JR
[1st_FILE.json](https://github.com/WinMerge/winmerge/files/13816560/1st_FILE.json)
[2nd_FILE.json](https://github.com/WinMerge/winmerge/files/13816561/2nd_FILE.json)

**Comment by sdottaka** [comment, marked as answer]
The attachments appear to be valid JSON files. Therefore, the Replace plugin is not required. Please specify as below.

<del> QueryJSON ".Data.RootChunk.renderAreaSettings.areaParameters |= sort_by(.HandleId)" </del>

~~~
QueryJSON ".Data.RootChunk.renderAreaSettings.areaParameters |= sort_by(.HandleId)"
~~~

![image](https://github.com/WinMerge/winmerge/assets/98126/fb50f33f-3661-448b-bc59-6b68084ee205)

**Reply by jrchaves** [reply]
WOW Mr. Tanaka, that worked super fine! I deeply deeply thank you for your support and attention (truly!).

For some reason I had to change a little bit the end of the instruction from (difference was in a blank space):

> QueryJSON ".Data.RootChunk.renderAreaSettings.areaParameters|=sort_by(.HandleId)"

to:

> QueryJSON ".Data.RootChunk.renderAreaSettings.areaParameters|= sort_by(.HandleId)"

Super appreciated!

Have you a great 2024!

best regards,

JR

**Comment by jrchaves** [comment]
Mr. Sawanaka, here I am again ... :) 
The code above is working really fine ... but now I got a different situation where I have more chained HandleID. 

I already tried several times but all failed ... I would like that before the comparison, the order of the 2nd FILE be the same as the 1st FILE, I mean, moving (in the 2nd FILE) the corresponding objects (HandleIDs and other), allowing the comparison between them.

Is it possible to take a look? Although there are several lines, just specific parts are inverted (the majority is in the same order).

Best regards,

JR
[1st_FILE.json](https://github.com/WinMerge/winmerge/files/13821843/1st_FILE.json)
[2nd_FILE.json](https://github.com/WinMerge/winmerge/files/13821844/2nd_FILE.json)

**Reply by sdottaka** [reply]
The attached files had the same content as the previous files.
Are you sure you're attaching the wrong files?

**Reply by jrchaves** [reply]
ouch ... My apologies ... For some reason I mixed them. Now I have attached the correct files. Thank you Sawanaka.

best regards,

JR
[1st_FILE.json](https://github.com/WinMerge/winmerge/files/13823968/1st_FILE.json)
[2nd_FILE.json](https://github.com/WinMerge/winmerge/files/13823969/2nd_FILE.json)

**Reply by sdottaka** [reply]
I'm not sure if it's the result you expected, but how about sorting by $type like below?
However, since I couldn't pass the $ character properly, I removed it using the Replace plugin.
~~~
Replace "$" "" | QueryJSON ".Data.RootChunk.worldRenderSettings.areaParameters |= sort_by(.Data.type)"
~~~

**Comment by jrchaves** [comment]
WOWWWWWWWWWWWWW! That was exactly what I expected Sawanka! This will help me a lot (more than 3 hours trying to achieve that!) ... Again, super thanks! Very appreciated!

best regards,

JR

**Comment by jrchaves** [comment]
Hi Mr. Sawanaka, 

I guess it´s the last time I will bother you with the same subject. I have another situation that both instructions above could not solve the order comparison issue:

I have tried:

> QueryJSON ".Data.RootChunk.renderAreaSettings.areaParameters|= sort_by(.HandleId)"

and also:

> Replace "$" "" | QueryJSON ".Data.RootChunk.worldRenderSettings.areaParameters |= sort_by(.Data.type)"

The 1st instruction seemed to work but it didn´t (some parts still in a different order).

i have tried many alternatives but all failed. Could you kindly take a look on it, please?

best regards,

JR
[1st_FILE.json](https://github.com/WinMerge/winmerge/files/13828598/1st_FILE.json)
[2nd_FILE.json](https://github.com/WinMerge/winmerge/files/13828601/2nd_FILE.json)

**Comment by sdottaka** [comment]
To sort by $type, specify as follows.
~~~
Replace "$" "" | QueryJSON ".Data.RootChunk.renderAreaSettings.areaParameters |= sort_by(.Data.type)"
~~~

**Comment by jrchaves** [comment]
Oh Mr. Sawanaka, thank you so much. Now I understood the logics! 

Have a great day!

best regards

JR

**Comment by aronduby** [comment]
sorry to resurrect this, but it's exactly what I need except I can't seem to find how to get the QueryJSON plugin installed. I see a bunch of references to it existing, and I see it listed in the available plugins list, but available from where?! I'm looking at `C:\Program Files (x86)\WinMerge\MergePlugins` and don't see anything listed there. 

Any help would be much appreciated.

**Reply by aronduby** [reply]
🤦‍♂️ I just needed to update winmerge, embarrassing I didn't think of that before commenting. Leaving to save someone else the embarrassment in the future.

**Reply by aronduby** [reply]
also, if you just want to do all the properties this seems to work for me `QueryJSON "--sort-keys"`

**Reply by sdottaka** [reply]
Info: The actual definition of QueryJSON is listed in the Plugins.xml file located in C:\Program Files\WinMerge\MergePlugins.

**Comment by nirmalparmar** [comment]
Check if this j[son compare tool](https://www.jsonreader.com/json-compare) helps
